### ISSUE TITLE
Confusing `not trait` error message

### ISSUE THREAD

**Issue by klutzy** [rust-lang/rust#10361]
```Rust
use t::T;
mod t {
    pub trait T {}
}

struct S;

impl T for S {}

impl T {} // oops

fn main() {}
```

```
a.rs:8:5: 8:6 error: T is not a trait
a.rs:8 impl T for S {}
            ^
```

This should show better error message. If `trait T` is not in mod, rustc prints "duplicate definition of module `T`".

**Comment by steveklabnik** [comment]
This error is much better:

```
hello.rs:10:1: 10:10 error: inherent implementations are only allowed on types defined in the current module [E0257]
hello.rs:10 impl T {} // oops
            ^~~~~~~~~
hello.rs:1:5: 1:9 note: import from other module here
hello.rs:1 use t::T;
               ^~~~

```

I'm going to consider this one closed.
